unit_test/docker_container_test.py

```diff
@@ -21,23 +25,178 @@
     DEFAULT_IMAGE = "quay.io/pypa/manylinux2014_ppc64le:2020-05-17-2f8ac3b"
 elif pm == "s390x":
     DEFAULT_IMAGE = "quay.io/pypa/manylinux2014_s390x:2020-05-17-2f8ac3b"
+else:
+    DEFAULT_IMAGE = ""
+
+
+temp_test_dir = None
+
+
+@atexit.register
+def _cleanup_tempdir():
+    """
+    Cleans up any configuration written by :func:`basis_container_kwargs`.
+
+    For podman tests, the user is not given write permissions by default in new
+    directories. As a workaround chown them before trying to delete them.
+
+    It may be possible to handle this more cleanly in pytest itself, but using
+    atexit works well enough for now.
+    """
+    import stat
+
+    global temp_test_dir
+    if temp_test_dir is not None:
+        print(f"CLEANUP temp_test_dir = {temp_test_dir!r}")  # type: ignore[unreachable]
+        for r, ds, fs in os.walk(temp_test_dir.name):
+            for d in ds:
+                dpath = os.path.join(r, d)
+                if not os.path.islink(dpath):
+                    perms = os.lstat(dpath).st_mode
+                    try:
+                        os.chmod(dpath, stat.S_IWUSR | perms)
+                    except Exception as ex:
+                        print(f"issue with dpath = {dpath!r}, {ex!r}")
+
+            for f in fs:
+                fpath = os.path.join(r, f)
+                if not os.path.islink(fpath):
+                    perms = os.lstat(fpath).st_mode
+                    try:
+                        os.chmod(fpath, stat.S_IWUSR | perms)
+                    except Exception as ex:
+                        print(f"issue with fpath = {fpath!r}, {ex!r}")
+                    else:
+                        os.unlink(fpath)
+        try:
+            temp_test_dir.cleanup()
+        except Exception as ex:
+            print(f"Issue cleaning up ex = {ex!r}")
+    temp_test_dir = None
+
+
+def basis_container_kwargs():
+    """
+    Generate keyword args that can be passed to to :class:`DockerContainer`.
+
+    This is used with :func:`pytest.mark.parametrize` to run each test with
+    different configuraions of each supported containers engine.
+
+    For docker we test the default configuration.
+
+    For podman we test the default configuration and a configuration with VFS
+    (virtual file system) enabled as the storage driver.
+    """
+
+    global temp_test_dir
+    if temp_test_dir is None:
+        # Only setup the temp directory once for all tests
+        temp_test_dir = tempfile.TemporaryDirectory(prefix="cibw_test_")
+
+    HAVE_DOCKER = bool(shutil.which("docker"))
+    HAVE_PODMAN = bool(shutil.which("podman"))
```

Comment:
These tests rely on the runtime detection of something, meaning that they might stop being run one day if podman/docker isn't found, and we wouldn't know, the tests would still be green.

I'd propose moving to a test command line flag, like the existing `--run-docker` command, so adding a `--run-podman` flag that would enable the running of these tests with podman. I set up the existing `--run-docker` flag, I'd be happy to help with this if you like?

Reply:
This is a good point; an oversight on my part. I'd be glad to have the help setting up `--run-podman`.

Reply:
Sure, I'll take a look.

Reply:
I see the logic in conftest.py to add the "--run-docker" flag. If there is a corresponding podman mark, is there a way to apply specific marks to different instances of the `pytest.mark.parametrize`? The current way I have it setup is that I parameterize the same test over different configuration of the OCI backend. I'm not sure how to get some of them to trigger and others not to.

Reply:
This is done, actually it ended up quite a lot neater using pytest 'parametrized fixtures'